WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift

```diff
@@ -0,0 +1,445 @@
+    ///
+    private lazy var statusResultsController: ResultsController<StorageOrderStatus> = {
+        let storageManager = AppDelegate.shared.storageManager
+        let predicate = NSPredicate(format: "siteID == %lld", StoresManager.shared.sessionManager.defaultStoreID ?? Int.min)
+        let descriptor = NSSortDescriptor(key: "slug", ascending: true)
+
+        return ResultsController<StorageOrderStatus>(storageManager: storageManager, matching: predicate, sortedBy: [descriptor])
+    }()
+
+    /// Products from an Order
+    ///
+    var products: [Product] {
+        return productResultsController.fetchedObjects
+    }
+
+    /// Indicates if we consider the shipment tracking plugin as reachable
+    /// https://github.com/woocommerce/woocommerce-ios/issues/852#issuecomment-482308373
+    ///
+    var trackingIsReachable: Bool = false {
+        didSet {
+            dataSource.trackingIsReachable = trackingIsReachable
+        }
+    }
+
+    var displaysBillingDetails: Bool {
+        set {
+            dataSource.displaysBillingDetails = newValue
+        }
+        get {
+            return dataSource.displaysBillingDetails
+        }
+    }
+
+    /// The datasource that will be used to render the Order Details screen
+    ///
+    private(set) lazy var dataSource: OrderDetailsDataSource = {
+        return OrderDetailsDataSource(order: self.order)
+    }()
+
+    /// Order Notes
+    ///
+    var orderNotes: [OrderNote] = [] {
+        didSet {
+            dataSource.orderNotes = orderNotes
+            onUIReloadRequired?()
+        }
+    }
+
+    /// Closure to be executed when the UI needs to be reloaded.
+    /// That could happen, for example, when new incoming data is detected
+    ///
+    var onUIReloadRequired: (() -> Void)?
+
+    /// Closure to be executed when a cell triggers an action
+    ///
+    var onCellAction: ((OrderDetailsDataSource.CellActionType, IndexPath?) -> Void)? {
+        didSet {
+            dataSource.onCellAction = onCellAction
+        }
+    }
+
+    /// Helper
+    ///
+    private let emailComposer = OrderEmailComposer()
+
+    private let messageComposer = OrderMessageComposer()
+
+    func lookUpOrderStatus(for order: Order) -> OrderStatus? {
+        return dataSource.lookUpOrderStatus(for: order)
+    }
+
+    func lookUpProduct(by productID: Int) -> Product? {
+        return dataSource.lookUpProduct(by: productID)
+    }
+}
+
+
+extension OrderDetailsViewModel {
+    func configureResultsControllers(onReload: @escaping () -> Void) {
+        configureTrackingResultsController(onReload: onReload)
+        configureProductResultsController(onReload: onReload)
+        configureStatusResultsController()
+    }
+
+    private func configureTrackingResultsController(onReload: @escaping () -> Void) {
+        trackingResultsController.onDidChangeContent = { [weak self] in
```

Comment:
I am not sure I follow @jaclync could you elaborate please?

Reply:
like
```
class ResultsController<T: ResultsControllerMutableType> {
  ...
  public var onDidChangeContent: ((ResultsController) -> Void)?
}
```
so that this use case could be:
```
trackingResultsController.onDidChangeContent = { [weak self] resultsController in
  self?.dataSource.orderTracking = resultsController.fetchedObjects ?? []
  ...
}
```
just a slight prevent from calling the incorrect results controller, but pretty minor